xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -1566,3 +1567,448 @@ func (s) TestRingHash_ReattemptWhenGoingFromTransientFailureToIdle(t *testing.T)
+		p := peer.Peer{}
+		_, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&p))
+		if errors.Is(err, context.DeadlineExceeded) {
+			t.Fatalf("Timed out waiting for rpc EmptyCall() to be routed to the expected backend")
+		}
+		peerAddr = p.Addr.String()
+	}
+}
+
+// Tests that when all backends are down, we keep reattempting.
+func (s) TestRingHash_ReattemptWhenAllEndpointsUnreachable(t *testing.T) {
+	lis, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatalf("Failed to create listener: %v", err)
+	}
+	restartableListener := testutils.NewRestartableListener(lis)
+	restartableServer := stubserver.StartTestService(t, &stubserver.StubServer{
+		Listener: restartableListener,
+		EmptyCallF: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			return &testpb.Empty{}, nil
+		},
+	})
+	defer restartableServer.Stop()
+
+	const clusterName = "cluster"
+	endpoints := endpointResource(t, clusterName, []string{restartableServer.Address})
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		ServiceName: clusterName,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+	route := headerHashRoute("new_route", virtualHostName, clusterName, "address_hash")
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	xdsServer, nodeID, xdsResolver := setupManagementServerAndResolver(t)
+	if err := xdsServer.Update(ctx, xdsUpdateOpts(nodeID, endpoints, cluster, route, listener)); err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	dopts := []grpc.DialOption{
+		grpc.WithResolvers(xdsResolver),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithConnectParams(fastConnectParams),
+	}
+	conn, err := grpc.NewClient("xds:///test.server", dopts...)
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	if got, want := conn.GetState(), connectivity.Idle; got != want {
+		t.Errorf("conn.GetState(): got %v, want %v", got, want)
+	}
+
+	t.Log("Stopping the backend server")
+	restartableListener.Stop()
+
+	_, err = client.EmptyCall(ctx, &testpb.Empty{})
+	if err == nil || status.Code(err) != codes.Unavailable {
+		t.Errorf("rpc EmptyCall() succeeded, want Unavailable error")
+	}
+
+	// Wait for channel to fail.
+	testutils.AwaitState(ctx, t, conn, connectivity.TransientFailure)
+
+	t.Log("Restarting the backend server")
+	restartableListener.Restart()
+
+	// Wait for channel to become connected without any pending RPC.
+	testutils.AwaitState(ctx, t, conn, connectivity.Ready)
+}
+
+// Tests that when a backend goes down, we will move on to the next subchannel
+// (with a lower priority).  When the backend comes back up, traffic will move
+// back.
+func (s) TestRingHash_SwitchToLowerPriorityAndThenBack(t *testing.T) {
```

Comment:
This test is interesting. This is basically testing `ring_hash`'s interaction with the `priority` LB policy right?

Reply:
Yes, and as this issue demonstrates, being able to move back up a priority is an important property of the balancer.